url_shortener/cmd/httpload/main.go

```diff
@@ -0,0 +1,291 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"sort"
+	"strings"
+	"time"
+)
+
+const (
+	exitCodeOk    = 0
+	exitCodeError = 1
+
+	nanoSecondRatio = 1.0e-9
+
+	usageMessage = `Usage: httpload [options...] URL
+
+	Options:
+		-w int		number of concurrent workers running (default: 50)
+		-n int		number of requests to run (default: 200): must be equal or greater than -w
+		-z duration	application duration to send requests (default: unlimited)
+`
+)
+
+type results []result
+
+type result struct {
+	status int
+	timing int64
+}
+
+var (
+	nWorkers    = flag.Int("w", 50, "number of concurrent workers running (default: 50)")
+	nRequests   = flag.Int("n", 200, "number of requests to run (default: 200)")
+	appDuration = flag.Duration("z", 0, "application duration to send requests (default: unlimited)")
+)
+
+func dumpStatuses(results results) string {
+	dumpBuilder := &strings.Builder{}
+
+	statuses := make(map[int]int)
+
+	for _, result := range results {
+		statuses[result.status]++
+	}
+
+	keys := make([]int, 0, len(statuses))
+
+	for key := range statuses {
+		keys = append(keys, key)
+	}
+
+	sort.Ints(keys)
+
+	for _, key := range keys {
+		code := float64(key)
+		count := float64(statuses[key])
+		fmt.Fprintf(dumpBuilder, "%s\n", fmt.Sprintf("[%3.f] %12.f response(s)", code, count))
+	}
+
+	return dumpBuilder.String()
+}
+
+func dumpTimings(results results) string {
+	dumpBuilder := &strings.Builder{}
+
+	totalTiming := float64(0)
+	slowestTiming := float64(0)
+	fastestTiming := float64(0)
+	averageTiming := float64(0)
+	requestsPerSecond := float64(0)
+
+	n := len(results)
+
+	if n > 0 {
```

Comment:
In that case I have to double the `Fprintf` statements. Or do you have an idea to print zero or one value?

Reply:
I'd say that in the case there are no results, it'd be preferable for `dumpTimings` to either do nothing or return an error.
Also, to keep the scope of the function as general as possible, it'd nice that dumpTimings takes an `io.Writer` in which you write the timings, rather than returning a string.

By doing so, you let to the caller the choice of the destination, it could be `os.Stdout`, `os.Stderr` or even a http.Response or whatever. 
I know the caller can achieve the same thing with a `string`, but trying to think in terms of streaming interfaces (`io.Writer`, `io.Reader`) is good as you're going to see a lot of that in Go